Various changes to OpenAL, especially to reach lower latencies.

LAGonauta opened the pull request:
Added checkbox to disable/enable audiostretching and to change the number of buffers.

Number of buffers and latency per buffer are user facing now, and the max latency is more predictable. (doesn't depend on OAL_MAX_SAMPLES)

Added X-RAM detection to force disable it on cards that have them. Even though the driver should disable it automatically for streaming sources it doesn't.

Removed "SoundSync.Wait()" on line 232 as we already guarantee there are samples available after it. Works a lot better without it, but I am not sure why it was needed. XAudio2 seems to also have it commented out.

Changed standard C-arrays to std::vector as they need to be resizable depending on the buffer size. Also I was running out of stack space, apparently...

Enabled AA filter on SoundTouch, quality seems to get a little bit better with it enabled and I couldn't see any CPU usage difference.

Misc formatting changes, not much.

Any suggestion is vastly welcome :)
If the changes are not going to be accepted I recommend at least removing the SoundSync.Wait() on line 232, I was really impressed when I first removed it.

**Needs to be tested on Linux and Mac.**

Also with my changes sometimes false-positive errors show-up when generating the buffers or the sources, even though it should not have been affected... I couldn't debug this, but if I also check for errors right before the "gen" call it doesn't do it. I even changed line 199/208 to better match OpenAL Soft examples, but not change...

For some reason WX is not disabling the checkboxes and the spincontrols when OpenAL is not selected... But I guess this might be a problem with WX3 or with my PC as it also happens on master here.

BhaaLseN commented:
Might've been easier to split up if you partitioned the commits a little finer ;)

Can you change those c-style casts to `static_cast` where appropriate (or `reinterpret_cast` if necessary)?
Also, we have a `Clamp` function somewhere that could replace the if/else MAX checks there.

@dolphin-emu-bot rebuild

lioncash commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -277,74 +316,112 @@ void OpenALStream::SoundLoop()
+      {
+        std::vector<float> dest(nSamples * STEREO_CHANNELS);
+        for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
+          dest.data()[i] = (float)realtimeBuffer.data()[i] / (1 << 15);
```
Why are all these indexing operations using data? `std::vector` already provides an `operator[]` for direct indexing.

LAGonauta replied:
No reason at all. Will change.

LAGonauta commented:
Just changed all casts to static_cast, removed useless .data() usage and I managed to make MathUtils::Clamp work.
Should also have fixed lint and the compilation on Linux.

ligfx commented:
Also, I'm in favor of removing SoundTouch entirely, if that's an viable option. It adds latency, complicates the code, and never sounds as good as what Dolphin does by default.

mbc07 commented:
AFAICT, SoundTouch provides the time stretching feature present on OpenAL, if removing it means losing time stretching it's a -1 for me. I mean, OpenAL isn't the default backend in neither of the platforms Dolphin runs, and the default backend (e.g. XAudio2) already have a very low latency, having a second backend which trades a little bit of the latency with time stretching is perfectly fine.

LAGonauta commented:
I hope it is better now :)
It probably won't pass lint, but every build should be working.

If (or when) it doesn't pass lint, can I just make another commit with everything I have to change, or do I need to redo the other commits?

JosJuice commented:
@dolphin-emu-bot rebuild

You need to redo the existing commits.

LAGonauta commented:
Hopefully I managed to fix all formatting issues... I really need to manage to install clang-format on Git Bash for Windows.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -247,8 +250,8 @@ void OpenALStream::SoundLoop()
     {
-      ALuint unqueuedBufferIds[OAL_MAX_BUFFERS];
-      alSourceUnqueueBuffers(uiSource, numBuffersProcessed, unqueuedBufferIds);
+      std::vector<ALuint> unqueuedBufferIds(OAL_MAX_BUFFERS);
```
This should be `std::array` in this commit. In the next commit, you can avoid the overhead of `std::vector` by using a `std::unique_ptr<ALuint[]>`.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -263,25 +266,25 @@ void OpenALStream::SoundLoop()
 
     // Convert the samples from short to float
-    float dest[OAL_MAX_SAMPLES * STEREO_CHANNELS];
+    std::vector<float> dest(OAL_MAX_SAMPLES * STEREO_CHANNELS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -295,15 +298,15 @@ void OpenALStream::SoundLoop()
     {
-      float dpl2[OAL_MAX_SAMPLES * OAL_MAX_BUFFERS * SURROUND_CHANNELS];
-      DPL2Decode(sampleBuffer, nSamples, dpl2);
+      std::vector<float> dpl2(OAL_MAX_SAMPLES * OAL_MAX_BUFFERS * SURROUND_CHANNELS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -316,46 +319,38 @@ void OpenALStream::SoundLoop()
       else if (fixed32_capable)
       {
-        int surround_int32[OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS];
+        std::vector<int> surround_int32(OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -316,46 +319,38 @@ void OpenALStream::SoundLoop()
-          else
-            surround_int32[i] = (int)dpl2[i];
+          dpl2[i] = dpl2[i] * (static_cast<s64>(1) << 31);
+		  surround_int32[i] = static_cast<s32>(MathUtil::Clamp(
```
You could avoid a lot of these casts by doing `surround_int32[i] = MathUtil::Clamp(dpl2[i], -1.0, 1.0) * INT32_MAX` (the minimum would be `INT32_MIN - 1` instead of the current `INT32_MIN`).

LAGonauta replied:
Just tested your suggestion, and sadly I have to cast dpl2 to double for it to work, like this:
`surround_int32[i] = static_cast<int>(MathUtil::Clamp(static_cast<double>(dpl2[i]), -1.0, 1.0) * INT32_MAX);`

So I guess it is matter of testing which is faster.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -316,46 +319,38 @@ void OpenALStream::SoundLoop()
       else
       {
-        short surround_short[OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS];
+        std::vector<short> surround_short(OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -316,46 +319,38 @@ void OpenALStream::SoundLoop()
-            surround_short[i] = SHRT_MIN;
-          else
-            surround_short[i] = (int)dpl2[i];
+		  surround_short[i] =
```
Similarly, you can do `surround_short[i] = MathUtil::Clamp(dpl2[i], -1.0, 1.0) * INT16_MAX`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -384,21 +379,21 @@ void OpenALStream::SoundLoop()
       {
         // Clamping is not necessary here, samples are always between (-1,1)
-        int stereo_int32[OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS];
+        std::vector<int> stereo_int32(OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -384,21 +379,21 @@ void OpenALStream::SoundLoop()
+        std::vector<int> stereo_int32(OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS);
         for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
-          stereo_int32[i] = (int)((float)sampleBuffer[i] * (INT64_C(1) << 31));
+          stereo_int32[i] = static_cast<int>(sampleBuffer[i] * (static_cast<s64>(1) << 31));
```
`* (static_cast<s64>(1) << 31)` can be replaced with `INT32_MAX`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -384,21 +379,21 @@ void OpenALStream::SoundLoop()
       {
         // Convert the samples from float to short
-        short stereo[OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS];
+        std::vector<short> stereo(OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS);
```
`std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -384,21 +379,21 @@ void OpenALStream::SoundLoop()
+        std::vector<short> stereo(OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS);
         for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
-          stereo[i] = (short)((float)sampleBuffer[i] * (1 << 15));
+          stereo[i] = static_cast<short>(sampleBuffer[i] * (1 << 15));
```
`* (1 << 15)` can be replaced with `* INT16_MAX`

ligfx commented on Source/Core/AudioCommon/OpenALStream.h:
```diff
@@ -89,9 +89,9 @@ class OpenALStream final : public SoundStream
-  ALuint uiBuffers[OAL_MAX_BUFFERS];
+  std::vector<short> realtimeBuffer;
+  std::vector<soundtouch::SAMPLETYPE> sampleBuffer;
+  std::vector<ALuint> uiBuffers;
```
These should all be `std::array`

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -178,13 +178,34 @@ void OpenALStream::SoundLoop()
-  numBuffers = SConfig::GetInstance().iLatency + 2;  // OpenAL requires a minimum of two buffers
+  u32 ul_frequency = m_mixer->GetSampleRate();
+  // OpenAL requires a minimum of two buffers
+  num_buffers = SConfig::GetInstance().iNumBuffers;
```
I'm still against the use of two user configuration variables when one would do fine. Can you peg num_buffers at 2 or 3?

LAGonauta replied:
As num_buffers can only be changed through the INI file I imagine it won't be a problem. Only those users that really want to change this value will be able to. I do not see a problem with having more choice options, but if needed I believe the best value for num_buffers is three.

Also I believe more buffers (instead of larger buffers) can help SoundTouch as it is possible to change the rate of each buffer more often, but I did not test this.

ligfx replied:
I don't understand. Why do you want to be able to change the number of buffers? The way I understand it, you generally want the minimum number of buffers that the hardware/software will support.

What you're saying about SoundTouch doesn't make sense to me. SoundTouch wants larger buffers, not more buffers, because its output size is adjusted based on its configuration. Larger buffers = higher quality shifting.

LAGonauta replied:
Makes sense. I will remove it then, I guess I would be the only person to experiment with it anyway.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -266,19 +287,21 @@ void OpenALStream::SoundLoop()
-    std::vector<float> dest(OAL_MAX_SAMPLES * STEREO_CHANNELS);
-    for (u32 i = 0; i < numSamples * STEREO_CHANNELS; ++i)
-      dest[i] = static_cast<float>(realtimeBuffer[i]) / (1 << 15);
+    std::vector<float> dest(rendered_samples * STEREO_CHANNELS);
```
Since `rendered_samples` will never be larger than `samplers_per_buffer`, these vectors should be allocated ahead of time.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -298,35 +321,36 @@ void OpenALStream::SoundLoop()
     }
 
-    unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer.data(), OAL_MAX_SAMPLES * numBuffers);
+	// We want SoundTouch to return already processed samples all at once
```
> We want SoundTouch to return already processed samples all at once

Why?

LAGonauta replied:
Samples seems to start to accumulate on an internal SoundTouch buffer if I use "rendered_samples" instead of "soundTouch.numSamples()", increasing the latency.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -298,35 +321,36 @@ void OpenALStream::SoundLoop()
     {
-      std::vector<float> dpl2(OAL_MAX_SAMPLES * OAL_MAX_BUFFERS * SURROUND_CHANNELS);
-      DPL2Decode(sampleBuffer.data(), nSamples, dpl2.data());
+      std::vector<float> dpl2(n_samples * SURROUND_CHANNELS);
```
Since we know the maximum size of the sample buffers ahead of time, this vector should be allocated ahead of time as well.

LAGonauta replied:
Sometimes n_samples > rendered_samples when using SoundTouch (I am not asking for SoundTouch a fixed number of samples, I am asking for all samples it has available). Also, as I said, if I change soundTouch.numSamples() back to rendered_samples the latency increases (and it is very noticeable)

I have a hunch that SoundTouch only likes to "send" samples which are a multiple of 576...

ligfx replied:
When you say the latency increases, do you mean it gets worse over time, or that it's constant but higher than configured for?

LAGonauta replied:
Constant but higher than configured before. I did not measure it, just heard it. 
The audio latency is definitively higher, annoyingly. I guess I can compare both by analyzing the number of unprocessed samples. Will do later.

ligfx replied:
Looks like SoundTouch has a way to query the size in frames of each output chunk, and the number of input frames it needs to produce each chunk. At a glance, the algorithm to calculate those looks relatively simple, so you might be able to adjust the SoundTouch parameters to match the buffer size.

LAGonauta replied:
I forgot about this comment before pushing my newest changes, I will look at it later.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -298,35 +321,36 @@ void OpenALStream::SoundLoop()
       else if (fixed32_capable)
       {
-        std::vector<int> surround_int32(OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS);
+        std::vector<int> surround_int32(n_samples * SURROUND_CHANNELS);
```
Should be allocated ahead of time

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -336,22 +360,22 @@ void OpenALStream::SoundLoop()
       else
       {
-        std::vector<short> surround_short(OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS);
+        std::vector<short> surround_short(n_samples * SURROUND_CHANNELS);
```
Should be allocated ahead of time.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -379,36 +403,36 @@ void OpenALStream::SoundLoop()
-        std::vector<int> stereo_int32(OAL_MAX_SAMPLES * STEREO_CHANNELS * OAL_MAX_BUFFERS);
-        for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
-          stereo_int32[i] = static_cast<int>(sampleBuffer[i] * (static_cast<s64>(1) << 31));
+        std::vector<int> stereo_int32(n_samples * STEREO_CHANNELS);
```
Should be allocated ahead of time.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -208,27 +229,27 @@ void OpenALStream::SoundLoop()
-  ALint iState = 0;
+  unsigned int next_buffer = 0;
+  unsigned int num_buffers_queued = 0;
+  ALint i_state = 0;
```
No need to stick to Hungarian notation, especially since it looks weird in underscore_case 😆

LAGonauta replied:
So, should I remove all hungarian notation from here?

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -86,13 +86,13 @@ void OpenALStream::Stop()
 
-  alSourceStop(uiSource);
-  alSourcei(uiSource, AL_BUFFER, 0);
+  alSourceStop(ui_source);
```
This commit needs to be split into "variable name changes" and "non-variable name changes." It's difficult to tell logic changes apart from syntactical cleanups at a glance.

ligfx replied:
Please split these apart.

LAGonauta replied:
Will do, I already changed all other commits and left this one last.

ligfx replied:
Okay, thanks!

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -264,7 +264,7 @@ void OpenALStream::SoundLoop()
     if (num_buffers == num_buffers_queued && !num_buffers_processed)
     {
-      soundSyncEvent.Wait();
+      // soundSyncEvent.Wait();
```
Might as well delete it.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -189,22 +190,22 @@ void OpenALStream::SoundLoop()
   if (SConfig::GetInstance().iLatency > 0)
   {
-	  samples_per_buffer = ul_frequency / 1000 * SConfig::GetInstance().iLatency / num_buffers;
+    samples_per_buffer = ul_frequency / 1000 * SConfig::GetInstance().iLatency / num_buffers;
```
Lint fixes from prior commits shouldn't be in this commit.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -314,23 +324,46 @@ void OpenALStream::SoundLoop()
+      }
+      else
+      {
+        alSourcef(ui_source, AL_PITCH, static_cast<ALfloat>(rate));
```
Doesn't this add latency as well? If users want pitch-shifting, they should use SoundTouch (it's probably better at it anyways).

LAGonauta replied:
It could add, sure. But OpenAL Soft is also well optimized for that. SoundTouch and OpenAL Soft's latency after pitch-shifting is minimal, and latency is probably nonexistent when using hardware based OpenAL drivers (which also hardware accelerates the effect, but I bet it is already so light on processor usage that it makes no difference).

ligfx replied:
This still doesn't make sense to me. Time-stretching and pitch-shifting are the same thing—if a user disables time-stretching, then it shouldn't just switch to a different implementation.

Does OpenAL actually do pitch-shifting, or does only resample? If it's resampling, the Mixer class already does that internally.

LAGonauta replied:
Sorry, maybe I am not understanding which is which.
I understand pitch-shifting as changing the tempo without correcting the pitch, when the emulation speed is slow the tempo is reduced and the pitch gets lower (gets shifted). This operation is fast.
Time-stretching I understand as changing the tempo and pitch-correcting the output, which is slow but outputs audio at the correct pitch.
If my understanding is wrong, what are the correct terms?

ligfx replied:
Yeah, the terminology is confusing. Pitch-shifting changes the pitch but not the tempo; time-stretching changes the tempo but not the pitch; and resampling/frequency-shifting changes both at the same time.

So the Mixer class does frequency-shifting internally (unless you ask it not to, with that boolean third parameter) and SoundTouch does time-stretching/pitch-shifting (which are the same: if you have a time-stretcher, you can combine that with a resampler to get a pitch-shifter, and vice-versa).

LAGonauta replied:
Now I understand! Just run some tests to compare both Dolphin's native frequency-shifting (cosider_framelimite = true), OpenAL Soft's and X-Fi's (consider_framelimit = false).

Dolphin's is OK, but probably was made to be fast and easy on the CPU. It doesn't shift the frequency very much (which is good) but causes "pockets" of no sound (stuttering?) (which is bad).
OpenAL Soft's and X-Fi's shifts the frequency a lot (which is bad) but the sound is smooth as butter.

I can record a video comparing all of them if you want.

I noticed no increase in latency, and the good thing about using OpenAL to frequency-shift is that we can update the pitch/playback-speed every time m_mixer->GetCurrentSpeed() changes, not only after a buffer is played. I admit that it won't change much as each buffer is really small (10ms or lower)...

ligfx replied:
Er, what do you mean? They should both be resampling by exactly the same amount. If you don't mind recording a video, that would be helpful.

LAGonauta replied:
Here they are:
Resampled using OpenAL:
https://youtu.be/3ldr1mumVdA

Resampled using Dolphin:
https://youtu.be/dBkP0nFBd9g

Time-stretching using SoundTouch:
https://youtu.be/0PVNWL3oPUQ

I also attached here an Audacity image showing what happens to the audio when resampled using Dolphin.
![captura2r](https://cloud.githubusercontent.com/assets/14236067/24731020/d8187b76-1a3d-11e7-9863-9ed63417505a.PNG)

ligfx replied:
Wow, yeah, that sounds terrible. I think something might be messed up in XAudio2, would you mind recording one for the OpenAL backend on the master branch?

LAGonauta replied:
Sure, but as the master branch always uses SoundTouch the sound is similar with the one I already recorded. XAudio2 from the master branch also sounds the same as the one I posted.
OpenAL on master branch:
https://youtu.be/V8AO9KDvUiQ

I also tested on OpenAL by disabling the OpenAL call to frequency-shift the audio but it sounds just like XAudio2... Maybe the resampling algorithm from Dolphin is not 100% mature yet? I guess I should test on Linux to check.

ligfx replied:
Oh right, urgh. So on macOS, at least, every backend sounds "buttery smooth," as you said. Not sure what's up with XAudio2.

ligfx replied:
@LAGonauta could you try building [ligfx/dolphin@openalnosoundtouch](https://github.com/ligfx/dolphin/tree/openalnosoundtouch), and let me know if that sounds correct?

LAGonauta replied:
Just tested it and sadly it sounds just like XAudio2. How does it sound on a Mac? The pitch gets lower just like my "Resampled using OpenAL" video?

ligfx replied:
Yes, it sounds normal on a Mac (though the optimal latency depends on what game is running / what else my computer is doing). I imagine this is probably because OpenAL doesn't do timer-based scheduling like the other backends, and, now that I look at it, XAudio2 doesn't seem to either...

ligfx commented on Source/Core/Core/ConfigManager.cpp:
```diff
@@ -248,6 +248,7 @@ void SConfig::SaveCoreSettings(IniFile& ini)
   core->Set("SelectedLanguage", SelectedLanguage);
   core->Set("OverrideGCLang", bOverrideGCLanguage);
   core->Set("DPL2Decoder", bDPL2Decoder);
+  core->Set("TimeStretch", bTimeStretch);
```
Ishiiruka uses `bTimeStretching`, it'd be nice to match.

LAGonauta replied:
Sure 👍

ligfx commented on Source/Core/DolphinWX/Config/AudioConfigPane.cpp:
```diff
@@ -55,50 +56,54 @@ void AudioConfigPane::InitializeGUI()
       _("Enables Dolby Pro Logic II emulation using 5.1 surround. Certain backends only."));
 
-  const int space5 = FromDIP(5);
+  m_soundtouch_timestretch_checkbox->SetToolTip(_("Enables timestretching support, smoother audio "
```
Does time-stretching really give smoother audio? My understanding is that Dolphin already produces smooth audio, and SoundTouch adds on pitch-shifting (albeit through time-stretching, but it's equivalent).

LAGonauta replied:
Sorry, my English is not perfect so I do not know how I would describe it. Any suggestion? Maybe "protects from buffer under-runs but can add latency"?

ligfx replied:
Maybe something like "Adjusts audio pitch when not running at full speed (adds latency)"?

ligfx commented on Source/Core/DolphinWX/Config/AudioConfigPane.cpp:
```diff
@@ -55,50 +56,54 @@ void AudioConfigPane::InitializeGUI()
-  const int space5 = FromDIP(5);
+  m_soundtouch_timestretch_checkbox->SetToolTip(_("Enables timestretching support, smoother audio "
+                                                  "but can add latency. Certain backends only."));
+  const int space6 = FromDIP(6);
```
Why the switch to `FromDIP(6)`? DolphinWX uses `FromDIP(5)` everywhere else.

LAGonauta replied:
This was my first time dealing with WX, so something is bound to be wrong... Reverted, as at first I thought it was causing compilation errors but now I see that it was another problem.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -178,13 +178,34 @@ void OpenALStream::SoundLoop()
+  }
+  else
+  {
+	  samples_per_buffer = ul_frequency / 1000 * 1 / num_buffers;
```
This is a latency of 1 second. The default should be something much smaller, like `const size_t BUFFER_SAMPLES = 512;  // ~10 ms - needs to be at least 240 for surround` from PulseAudioStream.cpp .

LAGonauta replied:
I might be doing something wrong then, am I?
For me if ul_frequency is 48000 and iLatency is 1 then it will have a full latency of 48 samples, which is 1 ms. Actually, I guess I should limit iLatency's minimum to 3 ms, not 1 ms...

ligfx replied:
Sorry, you're right, I dropped a couple zeroes! Yeah, a default 1 ms latency might be too fast for some users. I suggest making it the same as the PulseAudio backend.

ligfx commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -178,13 +178,34 @@ void OpenALStream::SoundLoop()
+  }
+
+  realtime_buffer.reserve(samples_per_buffer * STEREO_CHANNELS);
+  sample_buffer.reserve(samples_per_buffer * STEREO_CHANNELS * 20);
```
Why the `* 20`?

LAGonauta replied:
I guess I should comment about it on the code.
sample_buffer is the buffer for SoundTouch, which can stretch the number of samples up to ten times (as we set the minimum rate to 0.1). I am multiplying by 20 just to be safe.

LAGonauta commented:
Just pushed my changes to the branch. I hope I did not forget anything... Please let me know if I did.

LAGonauta commented:
PR not relevant anymore.
